### PR TITLE
Fix query failure when lambda is in UNNEST/VALUES

### LINKED ISSUE #9709: Lambda cannot be used inside unnest
```
presto> select * from unnest(transform(sequence(1, 5), x -> x * x));
Query 20180110_011726_00386_hgjdi failed: com.facebook.presto.sql.tree.ArithmeticBinaryExpression cannot be cast to java.lang.Long
java.lang.ClassCastException: com.facebook.presto.sql.tree.ArithmeticBinaryExpression cannot be cast to java.lang.Long
        at com.facebook.presto.$gen.VarArgsToMapAdapter_3751.varArgsToMap(Unknown Source)
        at java.lang.invoke.MethodHandleProxies$1.invoke(MethodHandleProxies.java:187)
        at com.sun.proxy.$Proxy457.apply(Unknown Source)
        at com.facebook.presto.$gen.ArrayTransform_3752.transform(Unknown Source)
        at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
        at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:649)
        at com.facebook.presto.sql.FunctionInvoker.invoke(FunctionInvoker.java:88)
        at com.facebook.presto.sql.planner.ExpressionInterpreter$Visitor.visitFunctionCall(ExpressionInterpreter.java:971)
        at com.facebook.presto.sql.tree.FunctionCall.accept(FunctionCall.java:114)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.planner.ExpressionInterpreter.evaluate(ExpressionInterpreter.java:271)
        at com.facebook.presto.sql.planner.ExpressionInterpreter.evaluateConstantExpression(ExpressionInterpreter.java:234)
        at com.facebook.presto.sql.planner.RelationPlanner.visitUnnest(RelationPlanner.java:523)
        at com.facebook.presto.sql.planner.RelationPlanner.visitUnnest(RelationPlanner.java:94)
        at com.facebook.presto.sql.tree.Unnest.accept(Unnest.java:62)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.planner.QueryPlanner.planFrom(QueryPlanner.java:269)
        at com.facebook.presto.sql.planner.QueryPlanner.plan(QueryPlanner.java:150)
        at com.facebook.presto.sql.planner.RelationPlanner.visitQuerySpecification(RelationPlanner.java:467)
        at com.facebook.presto.sql.planner.RelationPlanner.visitQuerySpecification(RelationPlanner.java:94)
        at com.facebook.presto.sql.tree.QuerySpecification.accept(QuerySpecification.java:127)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.planner.QueryPlanner.planQueryBody(QueryPlanner.java:258)
        at com.facebook.presto.sql.planner.QueryPlanner.plan(QueryPlanner.java:130)
        at com.facebook.presto.sql.planner.RelationPlanner.visitQuery(RelationPlanner.java:460)
        at com.facebook.presto.sql.planner.RelationPlanner.visitQuery(RelationPlanner.java:94)
        at com.facebook.presto.sql.tree.Query.accept(Query.java:94)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.planner.LogicalPlanner.createRelationPlan(LogicalPlanner.java:376)
        at com.facebook.presto.sql.planner.LogicalPlanner.planStatementWithoutOutput(LogicalPlanner.java:175)
        at com.facebook.presto.sql.planner.LogicalPlanner.planStatement(LogicalPlanner.java:156)
        at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:127)
        at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:122)
        at com.facebook.presto.execution.SqlQueryExecution.doAnalyzeQuery(SqlQueryExecution.java:363)
        at com.facebook.presto.execution.SqlQueryExecution.analyzeQuery(SqlQueryExecution.java:342)
        at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:274)
        at com.facebook.presto.execution.QueuedExecution.lambda$start$1(QueuedExecution.java:62)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

### PR DESCRIPTION
**Pull request by haozhun**
Fixes #9709

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 1]
```diff
@@ -444,7 +444,11 @@ else if (!javaType.isPrimitive()) {
         @Override
         protected Object visitIdentifier(Identifier node, Object context)
         {
-            return node;
+            // Identifier only exists before planning.
+            // ExpressionInterpreter should only be invoked after planning.
+            // As a result, this method should be unreachable.
+            // However, RelationPlanner.visitUnnest and visitValues invokes evaluateConstantExpression.
+            return ((SymbolResolver) context).getValue(new Symbol(node.getValue()));
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: is `context` always of type `SymbolResolver`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 2]
```diff
@@ -485,12 +485,16 @@ protected RelationPlan visitValues(Values node, Void context)
                 for (int i = 0; i < items.size(); i++) {
                     Expression expression = items.get(i);
                     expression = ExpressionTreeRewriter.rewriteWith(new ParameterRewriter(analysis.getParameters(), analysis), expression);
+                    // RelationPlanner should not invoke evaluateConstantExpression, which in turn invokes ExpressionInterpreter.
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: make this `TODO` (here & below)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 3]
```diff
@@ -444,7 +444,11 @@ else if (!javaType.isPrimitive()) {
         @Override
         protected Object visitIdentifier(Identifier node, Object context)
         {
-            return node;
+            // Identifier only exists before planning.
+            // ExpressionInterpreter should only be invoked after planning.
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Is this true? I thought the interpreter is invoked during planning for constant folding, etc.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java (hunk 1 above)
Comment (electrum) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Maybe add an `instanceof` check here and return null or throw an internal error?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 4]
```diff
@@ -485,12 +485,16 @@ protected RelationPlan visitValues(Values node, Void context)
...
+                    // This should happen in an optimizer.
                     Object constantValue = evaluateConstantExpression(expression, analysis.getCoercions(), metadata, session, analysis.getColumnReferences(), analysis.getParameters());
                     values.add(LiteralInterpreter.toExpression(constantValue, scope.getRelationType().getFieldByIndex(i).getType()));
                 }
             }
             else {
                 row = ExpressionTreeRewriter.rewriteWith(new ParameterRewriter(analysis.getParameters(), analysis), row);
+                // RelationPlanner should not invoke evaluateConstantExpression, which in turn invokes ExpressionInterpreter.
+                // This should happen in an optimizer.
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: I thought that this is happening in optimizer now.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java (hunk 2 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: it should go in separate commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java (hunk 4 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: See com.facebook.presto.sql.planner.iterative.rule.SimplifyExpressions#rewrite. I wonder why it needs to be here as well.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java (hunk 1 above)
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: No point. That would be a bug and it should fail in the same way every other checkState or assertion error does

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java (hunk 3 above)
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: No, it's not true. See "However" 2 lines down.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java (hunk 2 above)
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: This is closely related to the comments added in `ExpressionInterpreter.java`. I believe it's reasonable to be put together.

**Pull request merged**